File: sem1/hw4/task3/Stack.cpp
#include "Stack.h"

struct StackElement {
    char data;
    StackElement *next;
};

struct Stack {
    StackElement *top;
    int size;
};

Stack* createStack()
{
    Stack *s = new Stack;
    s->top = nullptr;
    s->size = 0;
    return s;

}

void push(Stack *s, char value)
{
    StackElement *newElement = new StackElement;
    newElement->data = value;
    newElement->next = s->top;
    s->top = newElement;
    s->size++;
}

bool isEmpty(Stack const *s)
{
    return s->size == 0;
}

char pop(Stack *s)
{
    if (isEmpty(s))
        return 'F';
    char x = s->top->data;
    StackElement *toDelete = s->top;
    s->top = s->top->next;
    delete toDelete;
    s->size--;
    return x;
}

void deleteStack(Stack *s)
{
    if (isEmpty(s))
    {
        delete s;
        return;
    }

    while (s->size != 0)
        pop(s);
    delete s;
}

File: sem2/hw1/task1/src/spbu/sem2/hw1/task1/Stack.java
package spbu.sem2.hw1.task1;

public class Stack<Type> {
    private class StackElement {
        public Type data;
        public StackElement next;
        public StackElement(Type data, StackElement next) {
            this.data = data;
            this.next = next;
        }
    }

    private StackElement top = null;
    private int size = 0;

    public void push (Type data) {
        StackElement toAdd = new StackElement(data, top);
        top = toAdd;
        size++;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public Type pop() {
        if (isEmpty())
            return null;
        size--;
        Type data = top.data;
        top = top.next;
        return data;
    }

   public int size() {
        return size;
   }

}
